pkg/edgeview/dispatcher/wss-server.go

```diff
@@ -0,0 +1,265 @@
+package main
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+type endPoint struct {
+	hostname  string
+	wsConn    *websocket.Conn
+}
+
+const (
+	noDeviceMsg string = "no device online\n+++Done+++"
+	tokenReqMsg string = "token is required"
+	moretwoMsg  string = "can't have more than 2 peers"
+)
+
+var upgrader = websocket.Upgrader{} // use default options
+// reqAddrTokeConn indexed by 'toke' then 'remoteAddr' strings
+var reqAddrTokenEP   map[string]map[string]endPoint
+// mutex for access the maps
+var connMutex sync.Mutex
+// connection id, keep inc
+var connID int
+// debug set
+var needDebug bool
+
+func socketHandler(w http.ResponseWriter, r *http.Request) {
+    // Upgrade our raw HTTP connection to a websocket based one
+    conn, err := upgrader.Upgrade(w, r, nil)
+    if err != nil {
+        fmt.Printf("Error during connection upgradation: %v\n", err)
+        return
+    }
+    defer conn.Close()
+
+	if _, ok := r.Header["X-Session-Token"]; !ok {
```

Comment:
This function needs indentation fixing.

Reply:
> Maybe we could also use edge-view to expose profiling data and the current stack trace of pillar microservices with the help of the [pprof package](https://pkg.go.dev/net/http/pprof) (pprof would be enabled in microservices end exposed via http if some config option is enabled). WDYT?

Good idea. I can add a story on this.

Reply:
@milan-zededa i realized we can do it right now, no story is needed. I just tried this:
1) compile gopprof into zedbox.go and listening on port 6060
2) with edge-view, do tcp/localhost:6060
3) another window on the laptop, get the file:

$ curl http://localhost:9001/debug/pprof/heap --output heap.file
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  100k    0  100k    0     0   363k      0 --:--:-- --:--:-- --:--:--  363k


$ go tool pprof heap.file
File: zedbox
Type: inuse_space
Time: Dec 10, 2021 at 1:51pm (PST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof)
(pprof) top
Showing nodes accounting for 8941.76kB, 61.33% of 14580.73kB total
Showing top 10 nodes out of 175
      flat  flat%   sum%        cum   cum%
 1722.14kB 11.81% 11.81%  1722.14kB 11.81%  github.com/google/gopacket/layers.init
 1542.33kB 10.58% 22.39%  1542.33kB 10.58%  encoding/json.Marshal
 1536.14kB 10.54% 32.92%  1536.14kB 10.54%  google.golang.org/protobuf/internal/impl.fieldInfoForScalar
 1025.44kB  7.03% 39.96%  1025.44kB  7.03%  encoding/pem.Decode
  526.13kB  3.61% 43.57%  1038.13kB  7.12%  sync.(*Map).LoadOrStore
  522.70kB  3.58% 47.15%   522.70kB  3.58%  regexp/syntax.(*compiler).inst
  522.06kB  3.58% 50.73%   522.06kB  3.58%  github.com/lf-edge/eve/pkg/pillar/cmd/nodeagent.parseSMARTData.func1
  515.19kB  3.53% 54.26%   515.19kB  3.53%  regexp.onePassCopy
     515kB  3.53% 57.80%      515kB  3.53%  google.golang.org/protobuf/internal/filedesc.(*File).initDecls
  514.63kB  3.53% 61.33%   514.63kB  3.53%  github.com/lf-edge/eve/pkg/pillar/types.(*ConfigItemSpecMap).AddIntItem

Reply:
> * compile gopprof into zedbox.go and listening on port 6060

@milan-zededa If we add this we should also add a rule to block port 6060 coming in on ethN ports.
@naiming-zededa do you automatically block any ports used between the edgeview container and EVE-OS or app instances so that they can not be accessed when coming in directly on ethN?

Reply:
From 'edge-view', no it does not install any ACLs. The above only shows if the EVE special image is coded to listen on this port (outside of edge-view domain), using edge-view 'tcp' option (it does not need to add anything for this gopprof usage), a remote laptop can access that gopprof service just like naturally routed into the device. But EVE can add any ACL rules when it offers this kind of service.

Reply:
But in this example of goprof that http server has no security what so ever so we'd have to block it.

Maybe we should have EVE-OS have ACLs to drop all inbound TCP/UDP on the physical ports unless there is a portmap for an application so that we don't accidentally open up holes which can be accessed from the LAN.

Reply:
Also in this particular case, the `pprof` is listening on 'localhost:6060', so only edge-view can get access to that besides ssh into the device. So this is ok.

Reply:
It would be good to check that the Linux kernel would block packets arriving on eth0 destined to 127.0.0.1:6060 (and ::1:6060 for IPv6). I suspect this is in some bogon-type check but we'd really depend on that being there for this pprof and other local port usage.

Reply:
@eriknordmark this is certainly the case, if the kernel does not listen on the ip:port combination, no ip service can be allowed in.

As in this case on my device:
the localhost:4822 is opened by 'edge-view' on my device:
07ab2c40-408a-4bc2-b9f2-8ca94235074f:~# netstat -ltnp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 10.1.0.1:53             0.0.0.0:*               LISTEN      3664/dnsmasq
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1156/sshd -D -e [li
tcp        0      0 0.0.0.0:4822            0.0.0.0:*               LISTEN      1237/guacd
tcp        0      0 127.0.0.1:8234          0.0.0.0:*               LISTEN      2108/edge-view

1) i can telnet to it if I'm on the device:
07ab2c40-408a-4bc2-b9f2-8ca94235074f:~# telnet 127.0.0.1:8234
Connected to 127.0.0.1:8234

2) but remotely to the eth0, will be denied:
07ab2c40-408a-4bc2-b9f2-8ca94235074f:~# ip -4 a
7: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 192.168.86.36/24 brd 192.168.86.255 scope global dynamic noprefixroute eth0
       valid_lft 73548sec preferred_lft 59083sec

% telnet 192.168.86.36:8234
192.168.86.36:8234: nodename nor servname provided, or not known

Reply:
I was referring to having an attacker which telnets to 127.0.0.1:8234 with a TCP/IP stack and ARP entry which sends the TCP SYN to the attacked node. Question is whether Linux drops packets arriving on non-lo0 destined to 127.0.0.1.

Reply:
I'm not sure how one can achieve such a task. But this issue is generic to Linux and not specific to EVE. As in kubernetes, the API server listens on 127.0.0.1:6443, and it's not reachable from external to control the apps inside unless some proxy service is enabled on the host to bridge that.

Reply:
one reference: https://books.google.com/books?id=_zhCEAAAQBAJ&pg=PT68&lpg=PT68&dq=does+linux+access+packet+from+external+to+127.0.0.1?&source=bl&ots=HYAfEmC63N&sig=ACfU3U0QKNoq3OOL-uypanEjyWGE6kL4PA&hl=en&sa=X&ved=2ahUKEwinl-68gPf3AhXTno4IHVVRC8sQ6AF6BAgbEAM#v=onepage&q=does%20linux%20access%20packet%20from%20external%20to%20127.0.0.1%3F&f=false

it says: "It is possible to spoof such a packet by writing a fake source address (i.e. spoofing that a packet has a source and destination address of 127.0.0.1). Linux will normally filter such a packet when it arrives at an external interface"